
<!DOCTYPE html
  PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
   <!--
This HTML was auto-generated from MATLAB code.
To make changes, update the MATLAB code and republish this document.
      --><title>Multivariate Data and Singular Value Decomposition</title><meta name="generator" content="MATLAB 9.6"><link rel="schema.DC" href="http://purl.org/dc/elements/1.1/"><meta name="DC.date" content="2019-07-21"><meta name="DC.source" content="problem_set1.m"><style type="text/css">
html,body,div,span,applet,object,iframe,h1,h2,h3,h4,h5,h6,p,blockquote,pre,a,abbr,acronym,address,big,cite,code,del,dfn,em,font,img,ins,kbd,q,s,samp,small,strike,strong,sub,sup,tt,var,b,u,i,center,dl,dt,dd,ol,ul,li,fieldset,form,label,legend,table,caption,tbody,tfoot,thead,tr,th,td{margin:0;padding:0;border:0;outline:0;font-size:100%;vertical-align:baseline;background:transparent}body{line-height:1}ol,ul{list-style:none}blockquote,q{quotes:none}blockquote:before,blockquote:after,q:before,q:after{content:'';content:none}:focus{outine:0}ins{text-decoration:none}del{text-decoration:line-through}table{border-collapse:collapse;border-spacing:0}

html { min-height:100%; margin-bottom:1px; }
html body { height:100%; margin:0px; font-family:Arial, Helvetica, sans-serif; font-size:10px; color:#000; line-height:140%; background:#fff none; overflow-y:scroll; }
html body td { vertical-align:top; text-align:left; }

h1 { padding:0px; margin:0px 0px 25px; font-family:Arial, Helvetica, sans-serif; font-size:1.5em; color:#d55000; line-height:100%; font-weight:normal; }
h2 { padding:0px; margin:0px 0px 8px; font-family:Arial, Helvetica, sans-serif; font-size:1.2em; color:#000; font-weight:bold; line-height:140%; border-bottom:1px solid #d6d4d4; display:block; }
h3 { padding:0px; margin:0px 0px 5px; font-family:Arial, Helvetica, sans-serif; font-size:1.1em; color:#000; font-weight:bold; line-height:140%; }

a { color:#005fce; text-decoration:none; }
a:hover { color:#005fce; text-decoration:underline; }
a:visited { color:#004aa0; text-decoration:none; }

p { padding:0px; margin:0px 0px 20px; }
img { padding:0px; margin:0px 0px 20px; border:none; }
p img, pre img, tt img, li img, h1 img, h2 img { margin-bottom:0px; } 

ul { padding:0px; margin:0px 0px 20px 23px; list-style:square; }
ul li { padding:0px; margin:0px 0px 7px 0px; }
ul li ul { padding:5px 0px 0px; margin:0px 0px 7px 23px; }
ul li ol li { list-style:decimal; }
ol { padding:0px; margin:0px 0px 20px 0px; list-style:decimal; }
ol li { padding:0px; margin:0px 0px 7px 23px; list-style-type:decimal; }
ol li ol { padding:5px 0px 0px; margin:0px 0px 7px 0px; }
ol li ol li { list-style-type:lower-alpha; }
ol li ul { padding-top:7px; }
ol li ul li { list-style:square; }

.content { font-size:1.2em; line-height:140%; padding: 20px; }

pre, code { font-size:12px; }
tt { font-size: 1.2em; }
pre { margin:0px 0px 20px; }
pre.codeinput { padding:10px; border:1px solid #d3d3d3; background:#f7f7f7; }
pre.codeoutput { padding:10px 11px; margin:0px 0px 20px; color:#4c4c4c; }
pre.error { color:red; }

@media print { pre.codeinput, pre.codeoutput { word-wrap:break-word; width:100%; } }

span.keyword { color:#0000FF }
span.comment { color:#228B22 }
span.string { color:#A020F0 }
span.untermstring { color:#B20000 }
span.syscmd { color:#B28C00 }

.footer { width:auto; padding:10px 0px; margin:25px 0px 0px; border-top:1px dotted #878787; font-size:0.8em; line-height:140%; font-style:italic; color:#878787; text-align:left; float:none; }
.footer p { margin:0px; }
.footer a { color:#878787; }
.footer a:hover { color:#878787; text-decoration:underline; }
.footer a:visited { color:#878787; }

table th { padding:7px 5px; text-align:left; vertical-align:middle; border: 1px solid #d6d4d4; font-weight:bold; }
table td { padding:7px 5px; text-align:left; vertical-align:top; border:1px solid #d6d4d4; }





  </style></head><body><div class="content"><h1>Multivariate Data and Singular Value Decomposition</h1><!--introduction--><p><b>*Dataset</b>*: Hi-C data obtained from human colon cells Comparing Healthy vs cells with Chr 7 trisomy</p><p><b>*Author</b>*: Koki Sasagawa</p><p><b>*Date</b>*: 2/11/19</p><p>Detailed explanation about Hi-C and biological concepts added to provide additional context</p><p>Useful context on HI-C data references from the following paper: "Three dimensional folding of chromosomes compartmentalizes the genome and can bring distant functional elements such as promoters and enhancers into close spatial proximity" - Berkum et al. 2010</p><p>How the data is generated: 1. Cells are fixed in vivo (take place in live cells) with chemicals such as formaldehyde (DNA-protein cross linkage) 2. Lyse (cut) the DNA using a restriction enzyme. Cross linked loci remain linked (Where the spatial location of chromosomes were touching) 3. Biotin incorporated as 5' overhangs (when double stranded DNA is cut, one strand usually ends up being longer, in this case the 5' end) are filled in (biotin is a biomolecular marker used to tag certain sequences and extract later with streptavidin which has a natural affinity to biotin) 4. Blunt-end ligation 5. Purify and shear DNA, pulldown biotin 6. Sequence paired-ends 7. Chromatin interactions are visualized as a heat map, where x and y axis represent loci in genomic order, and each pixel representing the number of observed interactions between. Each pixel represents interactions between two locus. Intensity corresponds to the number of reads (0-200 reads). The stronger the signal, the interaction occurred in a large fraction of the population. A weaker signal indicates the interaction occurred in a much smaller population.</p><!--/introduction--><h2>Contents</h2><div><ul><li><a href="#1">setup</a></li><li><a href="#2">Load data</a></li><li><a href="#3">1. Propose a method that can show the inequality between the two matrices using SVD</a></li><li><a href="#4">2. Compute the difference between chromosome 7 contact maps in the two data sets</a></li><li><a href="#5">Compute SVD and plot singular values</a></li><li><a href="#6">Compare Rank1 approximation</a></li><li><a href="#7">Compare difference between first principle component</a></li><li><a href="#8">3. Same computation for two other chromosomes of your choice</a></li><li><a href="#9">Compute SVD and plot singular values</a></li><li><a href="#10">Compare Rank1 approximation</a></li><li><a href="#11">Compare difference between first principle component</a></li><li><a href="#12">4. Compute the difference between the two data sets for the entire genome.</a></li><li><a href="#13">Compute SVD and plot singular values</a></li><li><a href="#14">Compare Rank1 approximation</a></li><li><a href="#15">Compare difference between first principle component</a></li></ul></div><h2 id="1">setup</h2><pre class="codeinput">clear
close <span class="string">all</span>
</pre><h2 id="2">Load data</h2><p>1. hcecHic (N x N matrix containing HCEC Hi-C sample, obs / exp normalized) 2. hcec7Hic (N x N matric containing HCEC+7 Hi-C sample, obs / exp normalized) 3. chrSart (23 x 1 vector containing starting locations of chromosomes in Hi-C matrices (e.g., chr 4 start at index 669)</p><pre class="codeinput">load(<span class="string">'hcec7hic1Mb.mat'</span>)
</pre><h2 id="3">1. Propose a method that can show the inequality between the two matrices using SVD</h2><p>We can compare the singularvalues of the two matrices which is obtained from computing SVD. Plot the singularvalues for the two matrices on a scatter plot.</p><p>Calculate the rank 1 approximation of matrices and compare the control with test.</p><p>Compute the first principal component of each matrix, then subtract the two PC1 vectors and see where they are most different.</p><h2 id="4">2. Compute the difference between chromosome 7 contact maps in the two data sets</h2><pre class="codeinput"><span class="comment">% First, plot the Hi-C data of chromosome 7</span>
fig(1) = figure(1);
set(fig(1), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:2
    ax(k) = subplot(1,2,k);
<span class="keyword">end</span>

<span class="comment">% select chr7 from normal cell data</span>
chrSelect = 7; <span class="comment">% select chr 7</span>
chrSelectIdx = chrStart(chrSelect): chrStart(chrSelect+1)-1; <span class="comment">% get the corresponding matrix indexes</span>
hcecHicChr7 = hcecHic(chrSelectIdx, chrSelectIdx);

subplot(ax(1));
imagesc(hcecHicChr7)
title(<span class="string">'Chr7 HCEC Control'</span>)
colorbar
axis(<span class="string">'square'</span>)

<span class="comment">% select chr7 from test cell data</span>
hcec7HicChr7 = hcec7Hic(chrSelectIdx, chrSelectIdx);

subplot(ax(2));
imagesc(hcec7HicChr7)
title(<span class="string">'Chr7 HCEC+7 TEST'</span>)
colorbar
axis(<span class="string">'square'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_01.png" alt=""> <h2 id="5">Compute SVD and plot singular values</h2><p>The '0' argument produces a economy-size decomposition by m-by-n matrix A 'economy size' decomposition removes extra rows or columns for zeros from the diagonal matrix of singular values, S, along with columns in either U or V that multiply those zeros in the expression A = U * S * V</p><pre class="codeinput"><span class="comment">% singular values for control</span>
[U1,S1,V1] = svd(hcecHicChr7,0);
sigma1 = diag(S1);

<span class="comment">% singular values for test</span>
[U2,S2,V2] = svd(hcec7HicChr7,0);
sigma2 = diag(S2);

<span class="comment">% scatterplot of top 4 largest singular values</span>
fig(2) = figure(2);
n = 1:4

p1 = scatter(n, sigma1(n), <span class="string">'r'</span>, <span class="string">'filled'</span>)
hold <span class="string">on</span> <span class="comment">% overlay second scatter plot on same figure</span>
p2 = scatter(n, sigma2(n), <span class="string">'b'</span>, <span class="string">'filled'</span>)
hold <span class="string">off</span>

legend([p1, p2], <span class="string">'hcec'</span>,<span class="string">'hcec+7'</span>, <span class="string">'FontSize'</span>, 14)
title(<span class="string">'Chr 7 singular values'</span>, <span class="string">'FontSize'</span>, 16)
</pre><pre class="codeoutput">
n =

     1     2     3     4


p1 = 

  Scatter with properties:

             Marker: 'o'
    MarkerEdgeColor: 'none'
    MarkerFaceColor: 'flat'
           SizeData: 36
          LineWidth: 0.5000
              XData: [1 2 3 4]
              YData: [167.3410 64.4796 23.8934 23.1735]
              ZData: [1&times;0 double]
              CData: [1 0 0]

  Use GET to show all properties


p2 = 

  Scatter with properties:

             Marker: 'o'
    MarkerEdgeColor: 'none'
    MarkerFaceColor: 'flat'
           SizeData: 36
          LineWidth: 0.5000
              XData: [1 2 3 4]
              YData: [175.8402 61.7937 28.8231 21.7912]
              ZData: [1&times;0 double]
              CData: [0 0 1]

  Use GET to show all properties

</pre><img vspace="5" hspace="5" src="problem_set1_02.png" alt=""> <h2 id="6">Compare Rank1 approximation</h2><pre class="codeinput">fig(2) = figure(2);
set(fig(2), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:2
    ax(k) = subplot(1,2,k);
<span class="keyword">end</span>

<span class="comment">% Control</span>
subplot(ax(1));
A1 = sigma1(1)*U1(:,1)*V1(:,1)';
imagesc(A1)
title(<span class="string">'Chr7 HCEC Control Rank 1 Approximation'</span>)
colorbar
axis(<span class="string">'square'</span>)

<span class="comment">% Test</span>
subplot(ax(2));
A2 = sigma2(1)*U2(:,1)*V2(:,1)';
imagesc(A2)
title(<span class="string">'Chr7 HCEC+7 TEST Rank 1 Approximation'</span>)
colorbar
axis(<span class="string">'square'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_03.png" alt=""> <h2 id="7">Compare difference between first principle component</h2><pre class="codeinput"><span class="comment">% PC1 of control</span>
<span class="comment">% mean center (on rows in this case since we want low dimensional)</span>
center = hcecHicChr7 - mean(hcecHicChr7,2);
[U,S,V] = svd(center',<span class="string">'econ'</span>);
svdScores = U * S;
PC1_control = svdScores(:,1);
x_values = 1:length(PC1_control);
pos_index_control = PC1_control &gt;= 0;
neg_index_control = PC1_control &lt; 0;

<span class="comment">% PC1 of test</span>
center = hcec7HicChr7 - mean(hcec7HicChr7,2);
[U,S,V] = svd(center',<span class="string">'econ'</span>);
svdScores = U * S;
PC1_test = svdScores(:,1);
pos_index_test = PC1_test &gt;= 0;
neg_index_test = PC1_test &lt; 0;

<span class="comment">% difference in PC1</span>
PC1_difference = PC1_test - PC1_control;
pos_index_difference = PC1_difference &gt;= 0;
neg_index_difference = PC1_difference &lt; 0;

fig(3) = figure(3);
set(fig(3), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:3
    ax(k) = subplot(3,1,k);
<span class="keyword">end</span>

<span class="comment">% make barplot</span>
subplot(ax(1));
bar(x_values(pos_index_control), PC1_control(pos_index_control), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_control), PC1_control(neg_index_control), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered'</span>)
title(<span class="string">'Chr7 HCEC'</span>)

subplot(ax(2));
bar(x_values(pos_index_test), PC1_test(pos_index_test), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_test), PC1_test(neg_index_test), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered'</span>)
title(<span class="string">'Chr7 HCEC+7'</span>)

subplot(ax(3));
bar(x_values(pos_index_difference), PC1_difference(pos_index_difference), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_difference), PC1_difference(neg_index_difference), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered difference'</span>)
title(<span class="string">'Chr7 HCEC+7 - HCEC'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_04.png" alt=""> <h2 id="8">3. Same computation for two other chromosomes of your choice</h2><p>Chromosome of interest: 20 Knockdown of TP53 and expression of K-RasV12 in +7 HCECs resulted in the emergence of trisomy 20, another nonrandom aneuploidy observed in 85% of Colorectal Cancer - https://www.sciencedirect.com/science/article/pii/S1476558611800163</p><pre class="codeinput"><span class="comment">% TP53 - provide instructions to create tumor protein p53 which acts as</span>
<span class="comment">% a tumor suppressor - regulate cell division.</span>
<span class="comment">% K-RasV12 - Group of protein callled K-Ras part of pathway known as the</span>
<span class="comment">% RAS/MAPK pathway - proteins that carry signals from outside of the cell</span>
<span class="comment">% to the cell nucleus. Turned on by GTP and off by GDP.</span>

fig(4) = figure(4);
set(fig(4), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:2
    ax(k) = subplot(1,2,k);
<span class="keyword">end</span>

<span class="comment">% Normal cell data</span>
chrSelect = 20;
chrSelectIdx = chrStart(chrSelect): chrStart(chrSelect+1)-1; <span class="comment">% get the corresponding matrix indexes</span>
hcecHicChr20 = hcecHic(chrSelectIdx, chrSelectIdx);

<span class="comment">% Test cell data</span>
hcec7HicChr20 = hcec7Hic(chrSelectIdx, chrSelectIdx);

subplot(ax(1));
imagesc(hcecHicChr20)
title(<span class="string">'Chr20 HCEC Control'</span>)
colorbar
axis(<span class="string">'square'</span>)

subplot(ax(2));
imagesc(hcec7HicChr20)
title(<span class="string">'Chr20 HCEC+7 TEST'</span>)
colorbar
axis(<span class="string">'square'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_05.png" alt=""> <h2 id="9">Compute SVD and plot singular values</h2><pre class="codeinput"><span class="comment">% singular values for control</span>
[U1,S1,V1] = svd(hcecHicChr20,0);
sigma1 = diag(S1);

<span class="comment">% singular values for test</span>
[U2,S2,V2] = svd(hcec7HicChr20,0);
sigma2 = diag(S2);

<span class="comment">% scatterplot of top 4 largest singular values</span>
fig(5) = figure(5);
n = 1:4

p1 = scatter(n, sigma1(n), <span class="string">'r'</span>, <span class="string">'filled'</span>)
hold <span class="string">on</span> <span class="comment">% overlay second scatter plot on same figure</span>
p2 = scatter(n, sigma2(n), <span class="string">'b'</span>, <span class="string">'filled'</span>)
hold <span class="string">off</span>

legend([p1, p2], <span class="string">'hcec'</span>,<span class="string">'hcec+7'</span>, <span class="string">'FontSize'</span>, 14)
title(<span class="string">'Chr 20 singular values'</span>, <span class="string">'FontSize'</span>, 16)
</pre><pre class="codeoutput">
n =

     1     2     3     4


p1 = 

  Scatter with properties:

             Marker: 'o'
    MarkerEdgeColor: 'none'
    MarkerFaceColor: 'flat'
           SizeData: 36
          LineWidth: 0.5000
              XData: [1 2 3 4]
              YData: [67.0326 26.1539 13.7053 10.8894]
              ZData: [1&times;0 double]
              CData: [1 0 0]

  Use GET to show all properties


p2 = 

  Scatter with properties:

             Marker: 'o'
    MarkerEdgeColor: 'none'
    MarkerFaceColor: 'flat'
           SizeData: 36
          LineWidth: 0.5000
              XData: [1 2 3 4]
              YData: [65.4634 23.6572 11.5294 10.8037]
              ZData: [1&times;0 double]
              CData: [0 0 1]

  Use GET to show all properties

</pre><img vspace="5" hspace="5" src="problem_set1_06.png" alt=""> <h2 id="10">Compare Rank1 approximation</h2><pre class="codeinput">fig(6) = figure(6);
set(fig(6), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:2
    ax(k) = subplot(1,2,k);
<span class="keyword">end</span>

<span class="comment">% Control</span>
subplot(ax(1));
A1 = sigma1(1)*U1(:,1)*V1(:,1)';
imagesc(A1)
title(<span class="string">'Chr20 HCEC Control Rank 1 Approximation'</span>)
colorbar
axis(<span class="string">'square'</span>)

<span class="comment">% Test</span>
subplot(ax(2));
A2 = sigma2(1)*U2(:,1)*V2(:,1)';
imagesc(A2)
title(<span class="string">'Chr20 HCEC+7 TEST Rank 1 Approximation'</span>)
colorbar
axis(<span class="string">'square'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_07.png" alt=""> <h2 id="11">Compare difference between first principle component</h2><pre class="codeinput"><span class="comment">% PC1 of control</span>
<span class="comment">% mean center (on rows in this case since we want low dimensional)</span>
center = hcecHicChr20 - mean(hcecHicChr20,2);
[U,S,V] = svd(center',<span class="string">'econ'</span>);
svdScores = U * S;
PC1_control = svdScores(:,1);
x_values = 1:length(PC1_control);
pos_index_control = PC1_control &gt;= 0;
neg_index_control = PC1_control &lt; 0;

<span class="comment">% PC1 of test</span>
center = hcec7HicChr20 - mean(hcec7HicChr20,2);
[U,S,V] = svd(center',<span class="string">'econ'</span>);
svdScores = U * S;
PC1_test = svdScores(:,1);
pos_index_test = PC1_test &gt;= 0;
neg_index_test = PC1_test &lt; 0;

<span class="comment">% difference in PC1</span>
PC1_difference = PC1_test - PC1_control;
pos_index_difference = PC1_difference &gt;= 0;
neg_index_difference = PC1_difference &lt; 0;

fig(7) = figure(7);
set(fig(7), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:3
    ax(k) = subplot(3,1,k);
<span class="keyword">end</span>

<span class="comment">% make barplot</span>
subplot(ax(1));
bar(x_values(pos_index_control), PC1_control(pos_index_control), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_control), PC1_control(neg_index_control), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered'</span>)
title(<span class="string">'Chr20 HCEC'</span>)

subplot(ax(2));
bar(x_values(pos_index_test), PC1_test(pos_index_test), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_test), PC1_test(neg_index_test), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered'</span>)
title(<span class="string">'Chr20 HCEC+7'</span>)

subplot(ax(3));
bar(x_values(pos_index_difference), PC1_difference(pos_index_difference), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_difference), PC1_difference(neg_index_difference), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered difference'</span>)
title(<span class="string">'Chr20 HCEC+7 - HCEC'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_08.png" alt=""> <h2 id="12">4. Compute the difference between the two data sets for the entire genome.</h2><pre class="codeinput">fig(8) = figure(8);
set(fig(8), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])
<span class="keyword">for</span> k = 1:2
    ax(k) = subplot(1,2,k);
<span class="keyword">end</span>

subplot(ax(1));
imagesc(hcecHic)
title(<span class="string">'HCEC Control'</span>)
colorbar
axis(<span class="string">'square'</span>)

subplot(ax(2));
imagesc(hcec7Hic)
title(<span class="string">'HCEC+7 TEST'</span>)
colorbar
axis(<span class="string">'square'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_09.png" alt=""> <h2 id="13">Compute SVD and plot singular values</h2><pre class="codeinput"><span class="comment">% singular values for control</span>
[U1,S1,V1] = svd(hcecHic,0);
sigma1 = diag(S1);

<span class="comment">% singular values for test</span>
[U2,S2,V2] = svd(hcec7Hic,0);
sigma2 = diag(S2);

<span class="comment">% scatterplot of top 4 largest singular values</span>
fig(9) = figure(9);
n = 1:4

p1 = scatter(n, sigma1(n), <span class="string">'r'</span>, <span class="string">'filled'</span>)
hold <span class="string">on</span> <span class="comment">% overlay second scatter plot on same figure</span>
p2 = scatter(n, sigma2(n), <span class="string">'b'</span>, <span class="string">'filled'</span>)
hold <span class="string">off</span>

legend([p1, p2], <span class="string">'hcec'</span>,<span class="string">'hcec+7'</span>, <span class="string">'FontSize'</span>, 14)
title(<span class="string">'Whole genome singular values'</span>, <span class="string">'FontSize'</span>, 16)
</pre><pre class="codeoutput">
n =

     1     2     3     4


p1 = 

  Scatter with properties:

             Marker: 'o'
    MarkerEdgeColor: 'none'
    MarkerFaceColor: 'flat'
           SizeData: 36
          LineWidth: 0.5000
              XData: [1 2 3 4]
              YData: [2.9014e+03 833.4139 422.7666 311.3927]
              ZData: [1&times;0 double]
              CData: [1 0 0]

  Use GET to show all properties


p2 = 

  Scatter with properties:

             Marker: 'o'
    MarkerEdgeColor: 'none'
    MarkerFaceColor: 'flat'
           SizeData: 36
          LineWidth: 0.5000
              XData: [1 2 3 4]
              YData: [2.9646e+03 788.2809 477.3763 259.7240]
              ZData: [1&times;0 double]
              CData: [0 0 1]

  Use GET to show all properties

</pre><img vspace="5" hspace="5" src="problem_set1_10.png" alt=""> <h2 id="14">Compare Rank1 approximation</h2><pre class="codeinput">fig(10) = figure(10);
set(fig(10), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:2
    ax(k) = subplot(1,2,k);
<span class="keyword">end</span>

<span class="comment">% Control</span>
subplot(ax(1));
A1 = sigma1(1)*U1(:,1)*V1(:,1)';
imagesc(A1)
title(<span class="string">'Whole genome HCEC Control Rank 1 Approximation'</span>)
colorbar
axis(<span class="string">'square'</span>)

<span class="comment">% Test</span>
subplot(ax(2));
A2 = sigma2(1)*U2(:,1)*V2(:,1)';
imagesc(A2)
title(<span class="string">'Whole genome HCEC+7 TEST Rank 1 Approximation'</span>)
colorbar
axis(<span class="string">'square'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_11.png" alt=""> <img vspace="5" hspace="5" src="problem_set1_12.png" alt=""> <h2 id="15">Compare difference between first principle component</h2><pre class="codeinput"><span class="comment">% PC1 of control</span>
<span class="comment">% mean center (on rows in this case since we want low dimensional)</span>
center = hcecHic - mean(hcecHic,2);
[U,S,V] = svd(center',<span class="string">'econ'</span>);
svdScores = U * S;
PC1_control = svdScores(:,1);
x_values = 1:length(PC1_control);
pos_index_control = PC1_control &gt;= 0;
neg_index_control = PC1_control &lt; 0;

<span class="comment">% PC1 of test</span>
center = hcec7Hic - mean(hcec7Hic,2);
[U,S,V] = svd(center',<span class="string">'econ'</span>);
svdScores = U * S;
PC1_test = svdScores(:,1);
pos_index_test = PC1_test &gt;= 0;
neg_index_test = PC1_test &lt; 0;

<span class="comment">% difference in PC1</span>
PC1_difference = PC1_test - PC1_control;
pos_index_difference = PC1_difference &gt;= 0;
neg_index_difference = PC1_difference &lt; 0;

fig(11) = figure(11);
set(fig(11), <span class="string">'unit'</span>, <span class="string">'normalized'</span>, <span class="string">'Position'</span>, [.15 .15 .7 .4])

<span class="keyword">for</span> k = 1:3
    ax(k) = subplot(3,1,k);
<span class="keyword">end</span>

<span class="comment">% make barplot</span>
subplot(ax(1));
bar(x_values(pos_index_control), PC1_control(pos_index_control), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_control), PC1_control(neg_index_control), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered'</span>)
title(<span class="string">'Whole genome HCEC'</span>)

subplot(ax(2));
bar(x_values(pos_index_test), PC1_test(pos_index_test), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_test), PC1_test(neg_index_test), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered'</span>)
title(<span class="string">'Whole genome HCEC+7'</span>)

subplot(ax(3));
bar(x_values(pos_index_difference), PC1_difference(pos_index_difference), <span class="string">'g'</span>, <span class="string">'BarWidth'</span>, 1)
hold <span class="string">on</span>
bar(x_values(neg_index_difference), PC1_difference(neg_index_difference), <span class="string">'b'</span>, <span class="string">'BarWidth'</span>, 1)
ylabel(<span class="string">'PC1 mean centered difference'</span>)
title(<span class="string">'Whole genome HCEC+7 - HCEC'</span>)
</pre><img vspace="5" hspace="5" src="problem_set1_13.png" alt=""> <p class="footer"><br><a href="https://www.mathworks.com/products/matlab/">Published with MATLAB&reg; R2019a</a><br></p></div><!--
##### SOURCE BEGIN #####
%% Multivariate Data and Singular Value Decomposition
% **Dataset**: Hi-C data obtained from human colon cells
% Comparing Healthy vs cells with Chr 7 trisomy
%
% **Author**: Koki Sasagawa
%
% **Date**: 2/11/19
%
% Detailed explanation about Hi-C and biological concepts added to provide
% additional context
%
% Useful context on HI-C data references from the following paper: 
% "Three dimensional folding of chromosomes compartmentalizes the genome and
% can bring distant functional elements such as promoters and enhancers
% into close spatial proximity" - Berkum et al. 2010
%
% How the data is generated:
% 1. Cells are fixed in vivo (take place in live cells) with chemicals such 
% as formaldehyde (DNA-protein cross linkage)
% 2. Lyse (cut) the DNA using a restriction enzyme. Cross linked loci remain
% linked (Where the spatial location of chromosomes were touching)
% 3. Biotin incorporated as 5' overhangs (when double stranded DNA is cut, 
% one strand usually ends up being longer, in this case the 5' end) are 
% filled in (biotin is a biomolecular marker used to tag certain sequences 
% and extract later with streptavidin which has a natural affinity to biotin) 
% 4. Blunt-end ligation 
% 5. Purify and shear DNA, pulldown biotin
% 6. Sequence paired-ends 
% 7. Chromatin interactions are visualized as a heat map, where x and y
% axis represent loci in genomic order, and each pixel representing the
% number of observed interactions between. Each pixel represents
% interactions between two locus. Intensity corresponds to the number of
% reads (0-200 reads). The stronger the signal, the interaction occurred in
% a large fraction of the population. A weaker signal indicates the
% interaction occurred in a much smaller population. 

%% setup
clear
close all

%% Load data
% 1. hcecHic (N x N matrix containing HCEC Hi-C sample, obs / exp normalized) 
% 2. hcec7Hic (N x N matric containing HCEC+7 Hi-C sample, obs / exp
% normalized) 
% 3. chrSart (23 x 1 vector containing starting locations of chromosomes in
% Hi-C matrices (e.g., chr 4 start at index 669)

load('hcec7hic1Mb.mat')
%% 1. Propose a method that can show the inequality between the two matrices using SVD
%
% We can compare the singularvalues of the two matrices which is obtained from
% computing SVD. Plot the singularvalues for the two matrices on a scatter plot. 
%
% Calculate the rank 1 approximation of matrices and compare the control
% with test. 
% 
% Compute the first principal component of each matrix, then subtract the
% two PC1 vectors and see where they are most different. 
%% 2. Compute the difference between chromosome 7 contact maps in the two data sets

% First, plot the Hi-C data of chromosome 7
fig(1) = figure(1);
set(fig(1), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:2
    ax(k) = subplot(1,2,k);
end

% select chr7 from normal cell data
chrSelect = 7; % select chr 7
chrSelectIdx = chrStart(chrSelect): chrStart(chrSelect+1)-1; % get the corresponding matrix indexes 
hcecHicChr7 = hcecHic(chrSelectIdx, chrSelectIdx); 

subplot(ax(1));
imagesc(hcecHicChr7)
title('Chr7 HCEC Control')
colorbar
axis('square')

% select chr7 from test cell data
hcec7HicChr7 = hcec7Hic(chrSelectIdx, chrSelectIdx);

subplot(ax(2));
imagesc(hcec7HicChr7) 
title('Chr7 HCEC+7 TEST')
colorbar
axis('square')
%% Compute SVD and plot singular values
% The '0' argument produces a economy-size decomposition by m-by-n matrix A
% 'economy size' decomposition removes extra rows or columns for zeros from
% the diagonal matrix of singular values, S, along with columns in either U
% or V that multiply those zeros in the expression A = U * S * V

% singular values for control
[U1,S1,V1] = svd(hcecHicChr7,0);
sigma1 = diag(S1); 

% singular values for test
[U2,S2,V2] = svd(hcec7HicChr7,0);
sigma2 = diag(S2); 

% scatterplot of top 4 largest singular values
fig(2) = figure(2);
n = 1:4 

p1 = scatter(n, sigma1(n), 'r', 'filled')
hold on % overlay second scatter plot on same figure
p2 = scatter(n, sigma2(n), 'b', 'filled')
hold off

legend([p1, p2], 'hcec','hcec+7', 'FontSize', 14)
title('Chr 7 singular values', 'FontSize', 16)

%% Compare Rank1 approximation
fig(2) = figure(2);
set(fig(2), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:2
    ax(k) = subplot(1,2,k);
end

% Control
subplot(ax(1));
A1 = sigma1(1)*U1(:,1)*V1(:,1)';
imagesc(A1)
title('Chr7 HCEC Control Rank 1 Approximation')
colorbar
axis('square')

% Test
subplot(ax(2));
A2 = sigma2(1)*U2(:,1)*V2(:,1)';
imagesc(A2)
title('Chr7 HCEC+7 TEST Rank 1 Approximation')
colorbar
axis('square')

%% Compare difference between first principle component

% PC1 of control
% mean center (on rows in this case since we want low dimensional)
center = hcecHicChr7 - mean(hcecHicChr7,2);
[U,S,V] = svd(center','econ');
svdScores = U * S;
PC1_control = svdScores(:,1);
x_values = 1:length(PC1_control);
pos_index_control = PC1_control >= 0;
neg_index_control = PC1_control < 0;

% PC1 of test
center = hcec7HicChr7 - mean(hcec7HicChr7,2);
[U,S,V] = svd(center','econ');
svdScores = U * S;
PC1_test = svdScores(:,1);
pos_index_test = PC1_test >= 0;
neg_index_test = PC1_test < 0;

% difference in PC1
PC1_difference = PC1_test - PC1_control;
pos_index_difference = PC1_difference >= 0;
neg_index_difference = PC1_difference < 0;

fig(3) = figure(3);
set(fig(3), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:3
    ax(k) = subplot(3,1,k);
end

% make barplot
subplot(ax(1));
bar(x_values(pos_index_control), PC1_control(pos_index_control), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_control), PC1_control(neg_index_control), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered')
title('Chr7 HCEC')

subplot(ax(2));
bar(x_values(pos_index_test), PC1_test(pos_index_test), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_test), PC1_test(neg_index_test), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered')
title('Chr7 HCEC+7')

subplot(ax(3));
bar(x_values(pos_index_difference), PC1_difference(pos_index_difference), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_difference), PC1_difference(neg_index_difference), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered difference')
title('Chr7 HCEC+7 - HCEC')
%% 3. Same computation for two other chromosomes of your choice
% Chromosome of interest: 20
% Knockdown of TP53 and expression of K-RasV12 in +7 HCECs resulted in the
% emergence of trisomy 20, another nonrandom aneuploidy observed in 85% of
% Colorectal Cancer - https://www.sciencedirect.com/science/article/pii/S1476558611800163

% TP53 - provide instructions to create tumor protein p53 which acts as
% a tumor suppressor - regulate cell division. 
% K-RasV12 - Group of protein callled K-Ras part of pathway known as the
% RAS/MAPK pathway - proteins that carry signals from outside of the cell
% to the cell nucleus. Turned on by GTP and off by GDP. 

fig(4) = figure(4);
set(fig(4), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:2
    ax(k) = subplot(1,2,k);
end

% Normal cell data
chrSelect = 20;
chrSelectIdx = chrStart(chrSelect): chrStart(chrSelect+1)-1; % get the corresponding matrix indexes 
hcecHicChr20 = hcecHic(chrSelectIdx, chrSelectIdx); 

% Test cell data
hcec7HicChr20 = hcec7Hic(chrSelectIdx, chrSelectIdx);

subplot(ax(1));
imagesc(hcecHicChr20)
title('Chr20 HCEC Control')
colorbar
axis('square')

subplot(ax(2));
imagesc(hcec7HicChr20) 
title('Chr20 HCEC+7 TEST')
colorbar
axis('square')

%% Compute SVD and plot singular values

% singular values for control
[U1,S1,V1] = svd(hcecHicChr20,0);
sigma1 = diag(S1); 

% singular values for test
[U2,S2,V2] = svd(hcec7HicChr20,0);
sigma2 = diag(S2); 

% scatterplot of top 4 largest singular values
fig(5) = figure(5);
n = 1:4

p1 = scatter(n, sigma1(n), 'r', 'filled')
hold on % overlay second scatter plot on same figure
p2 = scatter(n, sigma2(n), 'b', 'filled')
hold off

legend([p1, p2], 'hcec','hcec+7', 'FontSize', 14)
title('Chr 20 singular values', 'FontSize', 16)

%% Compare Rank1 approximation
fig(6) = figure(6);
set(fig(6), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:2
    ax(k) = subplot(1,2,k);
end

% Control
subplot(ax(1));
A1 = sigma1(1)*U1(:,1)*V1(:,1)';
imagesc(A1)
title('Chr20 HCEC Control Rank 1 Approximation')
colorbar
axis('square')

% Test
subplot(ax(2));
A2 = sigma2(1)*U2(:,1)*V2(:,1)';
imagesc(A2)
title('Chr20 HCEC+7 TEST Rank 1 Approximation')
colorbar
axis('square')


%% Compare difference between first principle component

% PC1 of control
% mean center (on rows in this case since we want low dimensional)
center = hcecHicChr20 - mean(hcecHicChr20,2);
[U,S,V] = svd(center','econ');
svdScores = U * S;
PC1_control = svdScores(:,1);
x_values = 1:length(PC1_control);
pos_index_control = PC1_control >= 0;
neg_index_control = PC1_control < 0;

% PC1 of test
center = hcec7HicChr20 - mean(hcec7HicChr20,2);
[U,S,V] = svd(center','econ');
svdScores = U * S;
PC1_test = svdScores(:,1);
pos_index_test = PC1_test >= 0;
neg_index_test = PC1_test < 0;

% difference in PC1
PC1_difference = PC1_test - PC1_control;
pos_index_difference = PC1_difference >= 0;
neg_index_difference = PC1_difference < 0;

fig(7) = figure(7);
set(fig(7), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:3
    ax(k) = subplot(3,1,k);
end

% make barplot
subplot(ax(1));
bar(x_values(pos_index_control), PC1_control(pos_index_control), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_control), PC1_control(neg_index_control), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered')
title('Chr20 HCEC')

subplot(ax(2));
bar(x_values(pos_index_test), PC1_test(pos_index_test), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_test), PC1_test(neg_index_test), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered')
title('Chr20 HCEC+7')

subplot(ax(3));
bar(x_values(pos_index_difference), PC1_difference(pos_index_difference), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_difference), PC1_difference(neg_index_difference), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered difference')
title('Chr20 HCEC+7 - HCEC')
%% 4. Compute the difference between the two data sets for the entire genome. 

fig(8) = figure(8);
set(fig(8), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])
for k = 1:2
    ax(k) = subplot(1,2,k);
end

subplot(ax(1));
imagesc(hcecHic)
title('HCEC Control')
colorbar
axis('square')

subplot(ax(2));
imagesc(hcec7Hic) 
title('HCEC+7 TEST')
colorbar
axis('square')

%% Compute SVD and plot singular values

% singular values for control
[U1,S1,V1] = svd(hcecHic,0);
sigma1 = diag(S1); 

% singular values for test
[U2,S2,V2] = svd(hcec7Hic,0);
sigma2 = diag(S2); 

% scatterplot of top 4 largest singular values
fig(9) = figure(9);
n = 1:4

p1 = scatter(n, sigma1(n), 'r', 'filled')
hold on % overlay second scatter plot on same figure
p2 = scatter(n, sigma2(n), 'b', 'filled')
hold off

legend([p1, p2], 'hcec','hcec+7', 'FontSize', 14)
title('Whole genome singular values', 'FontSize', 16)

%% Compare Rank1 approximation
fig(10) = figure(10);
set(fig(10), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:2
    ax(k) = subplot(1,2,k);
end

% Control
subplot(ax(1));
A1 = sigma1(1)*U1(:,1)*V1(:,1)';
imagesc(A1)
title('Whole genome HCEC Control Rank 1 Approximation')
colorbar
axis('square')

% Test
subplot(ax(2));
A2 = sigma2(1)*U2(:,1)*V2(:,1)';
imagesc(A2)
title('Whole genome HCEC+7 TEST Rank 1 Approximation')
colorbar
axis('square')

%% Compare difference between first principle component

% PC1 of control
% mean center (on rows in this case since we want low dimensional)
center = hcecHic - mean(hcecHic,2);
[U,S,V] = svd(center','econ');
svdScores = U * S;
PC1_control = svdScores(:,1);
x_values = 1:length(PC1_control);
pos_index_control = PC1_control >= 0;
neg_index_control = PC1_control < 0;

% PC1 of test
center = hcec7Hic - mean(hcec7Hic,2);
[U,S,V] = svd(center','econ');
svdScores = U * S;
PC1_test = svdScores(:,1);
pos_index_test = PC1_test >= 0;
neg_index_test = PC1_test < 0;

% difference in PC1
PC1_difference = PC1_test - PC1_control;
pos_index_difference = PC1_difference >= 0;
neg_index_difference = PC1_difference < 0;

fig(11) = figure(11);
set(fig(11), 'unit', 'normalized', 'Position', [.15 .15 .7 .4])

for k = 1:3
    ax(k) = subplot(3,1,k);
end

% make barplot
subplot(ax(1));
bar(x_values(pos_index_control), PC1_control(pos_index_control), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_control), PC1_control(neg_index_control), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered')
title('Whole genome HCEC')

subplot(ax(2));
bar(x_values(pos_index_test), PC1_test(pos_index_test), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_test), PC1_test(neg_index_test), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered')
title('Whole genome HCEC+7')

subplot(ax(3));
bar(x_values(pos_index_difference), PC1_difference(pos_index_difference), 'g', 'BarWidth', 1)
hold on
bar(x_values(neg_index_difference), PC1_difference(neg_index_difference), 'b', 'BarWidth', 1)
ylabel('PC1 mean centered difference')
title('Whole genome HCEC+7 - HCEC')

##### SOURCE END #####
--></body></html>